src/Moq/Interception/InterceptionAspects.cs

```diff
@@ -331,4 +332,117 @@ public static void Handle(Invocation invocation, Mock mock)
 			}
 		}
 	}
+
+	internal static class HandleAutoSetupProperties
+	{
+		private static readonly int AccessorPrefixLength = "?et_".Length; // get_ or set_
+
+		public static void Handle(Invocation invocation, Mock mock)
+		{
+			MethodInfo invocationMethod = invocation.Method;
+			if (invocationMethod.IsPropertyAccessor() && !invocationMethod.IsPropertyIndexerAccessor())
+			{
+				if (mock.Setups.FindMatchFor(invocation) == null)
+				{
+					PropertyInfo property = GetPropertyFromAccessorMethod(mock, invocationMethod, out Type propertyHolderType);
+					var expression = GetPropertyExpression(propertyHolderType, property);
+					var getter = property.GetGetMethod(true);
+
+					object value = null;
+					bool valueNotSet = true;
+
+					mock.Setups.Add(new AutoImplementedPropertyGetterSetup(expression, getter, () =>
+					{
+						if (valueNotSet)
+						{
+							object initialValue;
+							Mock innerMock;
+							try
+							{
+								initialValue = mock.GetDefaultValue(getter, out innerMock,
+									useAlternateProvider: mock.DefaultValueProvider);
+							}
+							catch
+							{
+								// Since this method performs a batch operation, a single failure of the default value
+								// provider should not tear down the whole operation. The empty default value provider
+								// is a safe fallback because it does not throw.
```

Comment:
Sorry, didn't get it... 
`SetupProperty` just uses `default(T)`, which I suspect equivalent to `DefaultValueProvider.Empty`:
https://github.com/moq/moq4/blob/9ea157cf61f1eb7deeadc419cac08d1aa9fd4f44/src/Moq/Mock.Generic.cs#L302-L325

But here we're using value provider, which can throw. Did you mean that we should allow value provider to throw and do not catch any exceptions?

Reply:
Apologies, I should have been more specific. I mentioned `SetupProperty` because it doesn't even try to guard against failure, but you're right, that's a bad comparison because unlike a default value provider, `default(T)` as used by `SetupProperty` cannot possibly fail in the first place.

Let me explain in a different fashion. The `catch` guard in `SetupAllProperties` is there to prevent one very specific situation: If your `mock` uses `DefaultValue.Mock` and you're calling `mock.SetupAllProperties()`, and `mock` has a property `P` of type `T` which is declared serializable but implements the serializable protocol incorrectly, then initializing `P` with a new `Mock<T>` instance can fail in DynamicProxy (which has logic to deal with serializable types). To avoid complete failure of `SetupAllProperties` because of that single property, there used to be [code](https://github.com/moq/moq4/blob/0094c40e13194f5ada1acd347f200ebeaf597050/Source/Mock.cs#L742-L757) that would essentially fall back to the `DefaultValue.Empty` (i.e. no attempt to mock `T`) in case of such failure. This logic has subsequently been refactored (by yours truly) into the `catch` clause you're seeing today.

Since `SetupAllProperties` is no longer a batch operation, faulty serializables can no longer cause the operation as a whole to fail; you'd have to access the `P` property, and I think it would be perfectly legitimate to let the DynamicProxy error about the faulty serializable type "bubble up" all the way to user code. (That would admittedly be a minute functional change that could annoy users every now and then.)

Any thoughts?

Reply:
Interesting, now I got it. Agree that catch can be removed. As for 
> Perhaps we need a new test for this aspect.

You meant that we should test that an exception from `DefaultValueProvider.GetDefaultValue` is bubbled up here?

Reply:
Ah, I was looking for that test, but couldn't remember what it was called nor where it was defined. :laughing:

> Will just change this test, since now it SHOULD throw

Perfect. The new test will essentially become a special case of the existing ` Accessing_property_of_bad_serializable_type_throws` test. Seems by throwing, Moq will become more consistent.